[[ RefactorPlatform ]] Restructuring

runrevmark opened the pull request:
This patch is the beginning of restructuring the platform implementation on Mac so that it can be made a separate loadable library.

The first step is to turn all the MCPlatform*Ref types into abstract interface + platform specific implementation style.

This is already the case for MCPlatformSurface, MCPlatformWindow, MCPlatformPlayer and MCPlatformSoundRecorder.

This patch adds MCPlatformWindowMask and MCPlatformSound to the list.

There is still MCPlatformMenu, MCPlatformCursor, MCPlatformColorTransform, MCPlatformLoadedFont and MCPlatformPrintSession to do.

Additionally, the clipboard and native layer abstractions need to be refactored in a similar way (on Mac - this should be possible by creating an engine side class which wraps a MCPlatform*Ref handle).

Once all platform refs are in this form, we can turn them into pure virtual public interface classes which the engine code uses directly (rather than using a procedural wrapping).

Once that is done, an MCPlatform class instance will have factory methods for each of the object types, and each object type can hold its own internal reference to its platform instance.

Beyond that, further work will be needed to thing or remove the methods the platform library requires from the engine.

runrevmark committed:
[[ PlatformRefactor ]] Make MCRegion and MCGRegion synonymous

This patch makes MCRegion a wrapper around MCGRegion meaning that
they can be used completely interchangeably.

TODO: Remove MCRegion entirely.

runrevmark committed:
[[ PlatformRefactor ]] Move initial path computation into engine

This patch moves the splitting of 'initial' into folder and file
into the engine as the process depends on engine logic, not
platform logic.

runrevmark committed:
[[ PlatformRefactor ]] Restructure MCPlatformWindowMask

This patch changes MCPlatformWindowMask to be based around an
partially abstract interface class so internal uses can use
directly rather than through a procedural API.

runrevmark committed:
[[ PlatformRefactor ]] Restructure MCPlatformSound

This patch turns MCPlatformSound into a virtual interface with
mac-specific implementation in a similar vein to MCPlatformWindowMask.

montegoulding committed:
[[ PlatformRefactor ]] Restructure MCPlatformLoadedFont

This patch turns MCPlatformLoadedFont into a virtual interface with
mac-specific implementation in a similar vein to MCPlatformWindowMask.

montegoulding commented:
GitHub apparently lets me push to your repo @runrevmark so I'll do that and we can keep this PR open.

montegoulding committed:
[[ PlatformRefactor ]] Restructure MCPlatformCursor

This patch turns MCPlatformCursor into a virtual interface with
mac-specific implementation in a similar vein to MCPlatformWindowMask

montegoulding committed:
[[ PlatformRefactor ]] Restructure MCPlatformColorTransform

This patch turns MCPlatformColorTransform into a virtual interface with
mac-specific implementation in a similar vein to MCPlatformWindowMask.

montegoulding committed:
[[ PlatformRefactor ]] Implement MCPlatformBase

This patch implements MCPlatformBase to provide common
methods for all platform API classes. At present this
is only reference counting.

montegoulding committed:
[[ PlatformRefactor ]] Restructure MCPlatformMenu

This patch turns MCPlatformMenu into a virtual interface with
mac-specific implementation in a similar vein to MCPlatformWindowMask

peter-b committed:
platform: Refactor "extern" Mac platform functions

- Move all `extern` Mac platform function declarations into a new
  `mac-extern.h` file.

- Modify `extern` Mac platform functions that create new instances
  to do so via `MCPlatform::makeRef()` and to return the new
  instance as a managed pointer.

- Fix up functions that use "create" functions to do explicitly
  extract the raw pointer with `Ref::unsafeTake()`.

peter-b committed:
Rename MCPlatform class to MCPlatformString

Prevent a name conflict between the `MCPlatform` namespace and the
`MCPlatform` class that implements `the platform` syntax.

peter-b committed:
platform: Move refactored platform API classes into MCPlatform namespace

Move "new-style" platform API classes into the `MCPlatform` namespace (e.g.
`MCPlatformSound` becomes `MCPlatform::Sound`.  Provide smart pointer
type aliases (e.g. `MCPlatform::SoundRef`) and compatibility aliases to
ensure that code that uses `MCPlatformSound` still works.

peter-b committed:
platform: Ensure MCPlatformLoadedFont is virtual

Remove fields from `MCPlatform::LoadedFont` so that it is virtual.

peter-b committed:
platform: Ensure MCPlatform::ColorTransform is fully virtual

Remove all members, constructors and destructors.

peter-b committed:
platform: Ensure that MCPlatform::Cursor is fully virtual

Remove all constructors, destructors and member variables.

peter-b committed:
platform: Ensure MCPlatform::Menu is virtual

Remove constructors and destructors from `MCPlatform::Menu`

peter-b committed:
platform: Ensure MCPlatform::Sound is fully virtual

Remove constructors and destructors from `MCPlatform::Sound`.

peter-b committed:
platform: Add MCPlatform::Base and MCPlatform::Ref

Add a new base reference-counted class implementation.  It:

- Handles overflow correctly by converting instances which overflow
  into permanent objects
- Allows `const` instances to have their reference counts increased &
  decreased

Add a managed pointer type for `MCPlatform::Base` pointers.  It
allows automatic conversion from `Y` to `T` where `T` is a base class
of `Y`.

Add a `makeRef` template function that allows you to ergonomically
construct platform interface object instances in return context, e.g.

    auto t_sound = makeRef<MCMacPlatformSound>();

or

    return makeRef<MCMacPlatformMenu>(args...)

peter-b committed:
platform: Switch to using MCPlatform::Base

Switch platform interface classes from `MCPlatformBase` to
`MCPlatform::Base` and remove the former.

montegoulding committed:
[Bug 19372] Fix crash opening multiple print dialogs

This patch retains and releases PMObjects appropriately from
the print dialog nest.

montegoulding commented:
@peter-b it appears `std::enable_if` is an issue. Any recommendations?

peter-b commented:
@montegoulding As long as it's building locally, don't worry about it for now.

montegoulding committed:
[[ PlatformRefactor ]] Restructure MCMacPlatformPrintDialogNest

This patch turns MCMacPlatformPrintDialogNest into a virtual interface
with mac-specific implementation in a similar vein to MCPlatformWindowMask.

montegoulding committed:
[[ PlatformRefactor ]] Refactor abort key thread

This patch refactors the abort key thread so that the static
variables are now members of the thread class and abort key
handling is via methods of the virtual MCPlatformCore interface.

montegoulding committed:
[[ PlatformRefactor ]] Move Init & Finalize calls to MCPlatformCore

montegoulding committed:
[[ Cleanup ]] Remove unused functions

montegoulding committed:
[[ PlatformRefactor ]] Move ShowMessageDialog to MCPlatformCore`

montegoulding committed:
[[ PlatformRefactor ]] Move some window methods to MCPlatformCore

This patch moves pseudo modal and moving window methods into
MCPlatformCore. Additionally CoreReference has been added to
the class heirarchy for classes that need a reference to
MCPlatformCore. In this patch MCPlatformWindow and MCPlatformMenu
begin inheriting from this class and are now created by
MCPlatformCore.

montegoulding committed:
[[ PlatformRefactor ]] Move menubar management into MCPlatformCore

montegoulding committed:
[[ PlatformRefactor ]] Move system property access to MCPlatformCore

montegoulding committed:
[[ PlatformRefactor ]] Move wait and callbacks to MCPlatformCore

This patch moves WaitForEvent, BreakWait, SheduleCallback and
platform player creation to MCPlatformCore. A MCiOSPlatformCore
object will need to be implemented to deal with cross compiled
classes needing to call these methods.

montegoulding committed:
[[ PlatformRefactor ]] Move snapshot to MCPlatformCore

montegoulding committed:
[[ PlatformRefactor ]] Move modal session to MCMacPlatformCore

montegoulding committed:
[[ PlatformRefactor ]] Move beep to MCPlatformCore

montegoulding committed:
[[ PlatformRefactor ]] Move death grip to MCPlatformCore

montegoulding committed:
[[ PlatformRefactor ]] Move FlushEvents to MCPlatformCore

montegoulding committed:
[[ PlatformRefactor ]] Move hide cursor to MCPlatformCore

montegoulding committed:
[[ PlatformRefactor ]] Move last mouse event to MCMacPlatformCore

montegoulding committed:
[[ PlatformRefactor ]] Move GetEventTime to MCPlatformCore

montegoulding committed:
[[ PlatformRefactor ]] Move GetWindowWithId to MCPlatformCore

montegoulding committed:
[[ PlatformRefactor ]] Move mouse handling to MCPlatformCore

montegoulding committed:
[[ PlatformRefactor ]] Move screen management to MCPlatformCore

montegoulding committed:
[[ PlatformRefactor ]] Constexpr default constructor MCMacPlatformCore

montegoulding committed:
[[ PlatformRefactor ]] Move backdrop to MCPlatformCore

montegoulding committed:
[[ PlatformRefactor ]] Move some static state to MCMacPlatformCore

montegoulding committed:
[[ PlatformRefactor ]] Move dis/enable screen updates to MCPlatformCore

montegoulding committed:
[[ PlatformRefactor ]] Move file & folder dialog to MCPlatformCore

montegoulding committed:
[[ Cleanup ]] Remove unused function declaration

montegoulding committed:
[[ PlatformRefactor ]] Remove static vars from color dialog

montegoulding committed:
[[ Cleanup ]] Remove unused extern declarations

montegoulding committed:
[[ PlatformRefactor ]] Move theme to MCPlatformCore

montegoulding committed:
[[ PlatformRefactor ]] Refactor script env

montegoulding committed:
[[ PlatformRefactor ]] Move remaining object creation to MCplatform

montegoulding committed:
[[ PlatformRefactor ]] Move death grip to correct header

montegoulding committed:
[[ PlatformRefactor ]] Add callback object

montegoulding committed:
[[ PlatformRefactor ]] Move cursor hidden to MCMacPlatformCore

montegoulding committed:
[[ PlatformRefactor ]] Move menu static vars to appropriate classes

montegoulding committed:
[[ PlatformRefactor ]] Remove uses of proceedural api from platform classes

montegoulding committed:
[[ PlatformRefactor ]] Add unreachable platform implementaion

montegoulding committed:
[[ PlatformRefactor ]] Add mac common core implementation for server

montegoulding committed:
[[ PlatformRefactor ]] Refactor native layer

This patch refactors the mac implementation of native layer to
operate on a MCPlatformNativeLayerRef.

montegoulding committed:
Merge branch 'develop' into refactor-platform_to_classes

This merge required a manual move of a patch to MCContainerView

montegoulding committed:
[[ PlatformRefactor ]] Use default constructors where feasible

montegoulding committed:
[[ PlatformRefactor ]] Refactor native theme and server support

This patch removes the unreachable core implementation and now
includes the entire platform API in server. Additionally native
theme drawing has been moved to MCPlatformCore.

montegoulding committed:
[[ PlatformRefactor ]] Add QueryInterface method for platform extensibility

montegoulding committed:
[[ PlatformRefactor ]] Move run block on main to MCPlatformCore

montegoulding committed:
[[ PlatformRefactor ]] Move some init/teardown back to platform_main

montegoulding committed:
[[ PlatformRefactpr ]] Remove redundant OS version check

montegoulding committed:
[[ PlatformRefactor ]] Remove proceedural MCPlatform call from lib

montegoulding committed:
[[ PlatformRefactor ]] Move some constructors and destructurs to headers

montegoulding committed:
[[ PlatformRefactor ]] Move MCPlatformWindow common methods into platform lib

montegoulding committed:
[[ PlatformRefactor ]] Remove use of MCscreen global

montegoulding committed:
[[ PlatformRefactor ]] Remove use of MCObject

montegoulding committed:
[[ PlatformRefactor ]] Remove use of font name globals

montegoulding committed:
[[ PlatormRefactor ]] Move proceedural platform funcs inline

montegoulding committed:
[[ PlatformRefactor ]] Add MinimalCore interface for non Mac platforms

montegoulding committed:
[[ PlatformRefactor ]] Rename mac-internal.h to mac-platform.h

montegoulding committed:
[[ PlatformRefactor ]] Move QueryInterface to MCPlatform::Base

montegoulding committed:
[[ PlatformRefactor ]] Use cross platform MCPlatformCreateCore

montegoulding committed:
[[ PlatformRefactor ]] Remove non-minimal methods

montegoulding committed:
[[ PlatformRefactor ]] Implement iOS platform core

montegoulding committed:
[[ PlatformRefactor ]] Implement android platform object

montegoulding committed:
Merge branch 'develop' into refactor-platform_to_classes

montegoulding committed:
[[ PlatformRefactor ]] Implement linux platform object

montegoulding committed:
[[ PlatformRefactor ]] Implement Windows platform object

montegoulding committed:
[[ PlatformRefactor ]] Implement emscripten platform object

montegoulding committed:
[[ PlatformRefactor ]] Include required headers for linux & android

montegoulding committed:
[[ PlatformRefactor ]] Move mac platform_main to separate file

montegoulding committed:
[[ PlatformRefactor ]] Fix copy/paste error

montegoulding committed:
[[ PlatformRefactor ]] Use class method instead of legacy function

montegoulding committed:
[[ PlatformRefactor ]] Use callback object for engine functions

This patch uses a the callback object to call on the engine that
created the platform objects. All dependencies are now resolved
with the exception of those in libfoundation and libgraphics
which will be in subsequent patches.

montegoulding committed:
[[ PlatformRefactor ]] Remove libgraphics dependency

runrevmark committed:
[[ Platform ]] Hooks into MCPlatformWaitForEvent

This patch adds some hooks to MCPlatformWaitForEvent and friends in lieu
of it being subclassable.

(cherry picked from commit 4a90aca4eb5999c661984e939e833d61d48a06b5)

Conflicts:
	engine/src/mac-abort.mm
	engine/src/mac-core.mm
	engine/src/platform.h

montegoulding committed:
[[ PlatformRefactor ]] Remove libfoundation dependency

This patch removes libfoundation as a dependency of libplatform.

montegoulding committed:
[[ PlatformRefactor ]] Add libplatform.gyp and build as static lib

montegoulding committed:
[[ Bug ]] Fix copy/paste error

montegoulding committed:
[[ Bug ]] Add missing virtual declaration

montegoulding committed:
[[ Bug ]] Undefine CreateWindow from windows.h

montegoulding committed:
Merge branch 'develop' into refactor-platform_to_classes

 Conflicts:
	engine/src/mac-abort.mm
	engine/src/mac-core.mm
	engine/src/platform.h

Keeping ours as conflicts were resolved in previous commit

montegoulding committed:
[[ Bug ]] Fixes for windows builds

montegoulding committed:
[[ PlatformRefactor ]] Make libplatform dependency of EXEs

montegoulding committed:
[[ Bug ]] Fix windows build issues

montegoulding committed:
[[ PlatformRefactor ]] Remove raw clipboard from minimal platform

montegoulding committed:
[[ PlatformRefactor ]] Callback for wait in minimal platform

montegoulding committed:
[[ PlatformRefactor ]] Add required callbacks for av-camera

montegoulding committed:
[[ Bug ]] Remove legacy #ifdef of CreatePlayer

montegoulding committed:
[[ PlatformRefactor ]] Remove raw clipboard from minimal platform

montegoulding committed:
[[ PlatformRefactor ]] Add to test dependencies for kernel

montegoulding committed:
[[ Cleanup ]] Remove unnecessary config

montegoulding committed:
[[ Bug ]] Don't use #import

montegoulding committed:
[[ Bug ]] Remove extern from namespace

montegoulding committed:
[[ PlatformRefactor ]] Fix 32 bit build issues

montegoulding committed:
[[ Config ]] Build emscripten on vulcan

montegoulding committed:
Merge branch 'develop' into refactor-platform_to_classes

The two main areas of conflict were changes to the platform
recorder API and changes to the mac native layer.
